fix to issue #14732

avishrivastava11 opened the pull request:
Added better documentation for evalf(subs=...)

#14732
Fixes #14732

I have added an explanation as to why we should prefer expr.evalf(subs=...) over expr.subs(...).evalf() or expr.evalf().subs(...)

Along with that, I've also added the explanation and the example given by @asmeurer in answering the question on stack-overflow ( https://stackoverflow.com/questions/50439037/what-exactly-is-the-subs-argument-of-the-evalf-method-of-sympy-expression-fo )

@asmeurer  @smichr  and other members and contributers
Pls review this PR, and let me know the improvements
Great working with you guys. :)

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

avishrivastava11 committed:
fix to issue #14732

sidhantnagpal commented on sympy/core/evalf.py:
```diff
@@ -1391,6 +1391,28 @@ def evalf(self, n=15, subs=None, maxn=100, chop=False, strict=False, quad=None,
+        ========
+
+        >>> import sympy
+        >>> x, y, z = sympy.symbols("x, y, z")
```
The definition of symbols is space separated: `symbols('x y z')`.
You can also directly use `>>> from sympy.abc import x, y, z`.

avishrivastava11 replied:
Done. @sidhantnagpal  Pls review the new PR. I have applied all the changes you suggested

sidhantnagpal commented on sympy/core/evalf.py:
```diff
@@ -1391,6 +1391,28 @@ def evalf(self, n=15, subs=None, maxn=100, chop=False, strict=False, quad=None,
+
+        >>> import sympy
+        >>> x, y, z = sympy.symbols("x, y, z")
+        >>> (x+y-z).subs({x:1e100,y:1,z:1e100}) #case_1
```
This should be `>>> (x + y - z).subs({x: 1e100, y: 1, z: 1e100}) # case_1` and similarly for others.
You can refer to the formatting being used currently.

avishrivastava11 replied:
Done. @sidhantnagpal  Pls review the new PR. I have applied all the changes you suggested

sidhantnagpal commented on sympy/core/evalf.py:
```diff
@@ -1391,6 +1391,28 @@ def evalf(self, n=15, subs=None, maxn=100, chop=False, strict=False, quad=None,
+        >>> (x+y-z).evalf(subs={x: 1e100, y: 1, z: 1e100}) #case_2
+        1.00000000000000
+
+        in case_1, naive substitution evaluates 1e100 + 1 - 1e100 which looses 1 because of
```
"In case ... which loses ..."

avishrivastava11 replied:
Done. @sidhantnagpal  Pls review the new PR. I have applied all the changes you suggested

avishrivastava11 committed:
modified PR #14870 by reviews from @sidhantnagpal

avishrivastava11 commented on sympy/core/evalf.py:
```diff
@@ -1391,6 +1391,28 @@ def evalf(self, n=15, subs=None, maxn=100, chop=False, strict=False, quad=None,
+        or expr.evalf().subs(...) since expr.evalf(subs=...) avoids the loss of
+        significance caused due to naive substitution as the expression is run
+        through the evalf algorithm which takes into account, the various issues
+        that can cause loss of significance. Naive substitution can also include
```
@sidhantnagpal 
`avoids the loss of significance caused due to naive substitution as the expression is run
through the evalf algorithm which takes into account, the various issues that can cause loss of significance`
this is a single sentence but here too I have used `loss of significance` as you said, twice just to convey the context. I don't think we should overdo it anymore.

avishrivastava11 commented on sympy/core/evalf.py:
```diff
@@ -1391,6 +1391,28 @@ def evalf(self, n=15, subs=None, maxn=100, chop=False, strict=False, quad=None,
             verbose=<bool>
                 Print debug information (default=False)
 
+        Note :
```
Done. @sidhantnagpal  Pls review the new PR. I have applied all the changes you suggested

avishrivastava11 commented:
@asmeurer and other members
Request you to review this PR and suggest changes.
If there are no more changes, pls merge this PR.
Thanks

avishrivastava11 committed:
modified evalf.py after @sidhantnagpal suggestion

avishrivastava11 commented on sympy/core/evalf.py:
```diff
@@ -1391,6 +1391,31 @@ def evalf(self, n=15, subs=None, maxn=100, chop=False, strict=False, quad=None,
                 Print debug information (default=False)
 
+        Notes
+        ========
```
Done
@sidhantnagpal 
@asmeurer and other members.
If there are no other changes, please merge this PR.
Thanks

avishrivastava11 commented:
@asmeurer , @jksuom , @certik can you review this PR and if there are no issues, please merge it.

smichr commented on sympy/core/evalf.py:
```diff
@@ -1391,6 +1391,31 @@ def evalf(self, n=15, subs=None, maxn=100, chop=False, strict=False, quad=None,
+
+        >>> import sympy
+        >>> from sympy.abc import x, y, z
+        >>> (x + y - z).subs({x: 1e100, y: 1, z: 1e100}) #case_1
```
->suggest the following for the discussion

```
Notes
=====

When using Floats in a computation, precision errors may affect the 
final result when the Float values are substituted into an expression. 
The way to obtain an accurate result is to pass the desired values as 
the subs argument when doing the evaluation. For example, adding 
1e16 (a Float) to 1 will truncate to 1e16 and then subtracting 1e16 
will give 0 -- that is the result obtained from the following when trying 
to evaluate by simple substitution of values into the expression:

>>> values = {x: 1e16, y: 1, z: 1e16}
>>> (x + y - z).subs(values)
0

If evalf is used and the desired values are sent as the 
argument for subs, the correct value is obtained:

>>> (x + y - z).evalf(subs=values)
1

Using the subs argument for evalf is the accurate way to 
evaluate an expression at a desired set of values when 
precision errors may otherwise affect the calculation.
```

avishrivastava11 replied:
Sure, I'll add this.

smichr replied:
I was suggesting replacement of what is there since it provides some additional specificity for why the subs argument should be used.

avishrivastava11 committed:
modified evalf.py according to changes suggested by @smichr

avishrivastava11 commented:
@smichr I have added the changes you suggested. Pls review and if there's nothing else to modify, pls merge.

avishrivastava11 committed:
corrected errors given by Travis CI in prev commit

avishrivastava11 commented:
@certik Review please

avishrivastava11 commented:
@jksuom @smichr  Review please.

smichr committed:
edit discussion about using subs in evalf

Hmmm...I wonder if subs could automatically call evalf if the results of the substitution would be a number.

smichr commented:
See what you think about the edits (you or anyone). If there are no objections, I will commit this.

smichr committed:
Update evalf.py

smichr committed:
Update evalf.py

smichr commented:
@avishrivastava11, thanks for getting this going. This was a needed addition to the docs. As I wrote in one of my commit notes, I wonder if this sort of evaluation should be done automatically when doing subs if there are no free symbols in the substitution result. That can be a new issue/topic for discussion, however.

asmeurer commented:
Oops, I forgot to include the logic so that it doesn't try to update the changelog when there is no entry. I've fixed the bug. You can ignore the error.
